### ISSUE TITLE
Smudgeplot interpretation Plukenetia volubilis

### ISSUE THREAD

**Issue by sivico26** [KamilSJaron/smudgeplot#36]
Hi Kamil, I have troubles understanding my smudgeplot. I have used the following command to generate it

```
ls $folder/reads/*.fastq > $folder/FILES
srun kmc -k21 -t$threads -m64 -ci1 -cs10000 @FILES kmer_counts tmp
srun kmc_tools -t$threads transform kmer_counts -cx10000 histogram kmer_k21.hist

L=$(kmer_cov_cutoff.R kmer_k21.hist L)
U=$(kmer_cov_cutoff.R kmer_k21.hist U)
echo $L $U # these need to be sane values like 30 800 or so
srun kmc_tools -t$threads transform kmer_counts -ci$L -cx$U dump -s kmer_k21.dump
srun hetkmers.py -k 21 -t $threads -o kmer_pairs < kmer_k21.dump
srun smudgeplot.R -i kmer_pairs_coverages_2.tsv -o sacha_ploidy -t "Plukenetia volubilis"
```

and it look like this:

![sacha_ploidy_smudgeplot](https://user-images.githubusercontent.com/37163025/57697458-8d824a00-7653-11e9-92f9-14a33ec54246.png)

![sacha_ploidy_smudgeplot_log10](https://user-images.githubusercontent.com/37163025/57697470-96731b80-7653-11e9-9935-b6b2261703f9.png)

Checking L and U, they were set to 10 and 4000 respectively.

Now, I have an indication from the literature that perennial plants in this family are hexaploid. So I am a little bit surprised that most of the k-mers are in the triploid smudge. how strong this suggests it is a triploid? 

Also, my understanding is that triploid organisms cannot reproduce sexually due to problems in chromosomes segregation during meiosis (here my basics could go really wrong), which contradicts the fact that some lab-mates have successfully done sexual crosses between several individuals.

Do you think it is possible that, if it is an homopolyploid, the chromosomes are so similar that it is not possible to differentiate the k-mers and, as a result, they got mixed in the same smudge?
In other words, could that triploid be actually a hexaploid with really similar chromosomes? I would like to know if I could be sure this result discards the literature hypothesis, or if it is still possible

How should I understand my smudgeplot?

Thanks in advance

**Comment by KamilSJaron** [comment]
Hello,

1. I see that the smudgeplot is misslabeled. AAB can not be AAB because of it's position on the x-axis, it's either AB or AABB, then AAAB must be either AAB or AAAABB. Sorry for the mess. On the brighter note, I think this should be already fixed already in the next version of smudgeplot (that I will release later this evening). Would you be willing to share me `*_coverages_2.tsv` and `kmer_k21.hist` files to fix the smudge labeling?

Could you please also run genomescope on `kmer_k21.hist`? We need to check what is the most likely 1n coverage estimate.

Whichever it will be, it's sort of strange this to be hexaploid. The closest smudgeplot I have generated was [strawberry](https://github.com/tbenavi1/smudgeplot/wiki/strawberry-tutorial) and there the pattern was very different.

**Comment by sivico26** [comment]
Hello Kamil,

Just for the record, I used version 0.1.3.

Here there are the files you requested: 

[seg_fault_examples.tar.gz](https://github.com/KamilSJaron/smudgeplot/files/3182460/seg_fault_examples.tar.gz)
[kmer_k21.hist.txt](https://github.com/KamilSJaron/smudgeplot/files/3182467/kmer_k21.hist.txt)

I shall give more information about my data: It is from WGS Illumina paired-end sequencing from leaf tissue, which means it included not only gDNA but mtDNA and cpDNA as well. I managed to assemble the plastids and filter the paired reads aligning to them. I keep the reads which were paired and did not align to any of the plastids as nuclear reads.

I used these "nuclear reads" to generate the smudgeplots I showed you above. Also, I used trimmomatic to feed KMC only with high-quality reads (just 6% were discarded though). 

I suspect the filtering is why the `kmer_k21.hist` is weird (_i.e._ there is no odd kmer counting). 

I also used genomescope, but I am a little bit skeptical about its output. I have run 3 analysis so far

1. [WGS reads -max kmer coverage 10000x](http://qb.cshl.edu/genomescope/analysis.php?code=zJshDsPHVQi02KyUKGht) (with the original data, without filtering)
2. [Only nuclear reads - max kmer coverage 10000x](http://qb.cshl.edu/genomescope/analysis.php?code=3d7V4Asa3wdz4zqZLyzW)
3. [Only nuclear reads - max kmer coverage 1000x](http://qb.cshl.edu/genomescope/analysis.php?code=VkfcOwCiB1Idk9odmKnI)

Also, some preliminary assemblies with [Flye](https://github.com/fenderglass/Flye/blob/flye/docs/USAGE.md) with some nanopore data I have from the same sample, and the genome size I found is ~284 Mb. 

So I am skeptical with genomescope due to the variability of its results, its estimates do not agree with my assemblies and they state the tool is developed for diploid genomes and polyploids could rather be off estimation. However, I am not that familiar with genomescope and kmer-analysis so any insights in this regard would be really appreciated. 

What do you think so far?

I will rerun the analysis with the new version and will return to you then.

**Comment by KamilSJaron** [comment]
Nice. We have a couple more pieces to the puzzle.

> assemble the plastids and filter the paired reads aligning to them. I keep the reads which were paired and did not align to any of the plastids as nuclear reads.

I think it's unnecessary to filter out mt/ct DNA (smudgeplot is quite robust to high freq kmers) and I also have noticed that the kmer historgram you have shared with me has 0 kmers covered by odd numbers, which is strange. Is it possble that there is something wrong in the filtering step? You expected all the coverages to be represeted, not just even numbers...

> I used trimmomatic to feed KMC only with high-quality reads (just 6% were discarded though).

Excelent, trimming always helps.

> I also used genomescope, but I am a little bit skeptical about its output.

Here we will care about the kmer distribution rather about the estimated values of heterozygosity and repeats, although we will use at least the genome size estimate. Since the Only nuclear genes is strange (with all the 0s), I will focus on WGS reads.

Smudgeplot has a very brigh smudge ~204x (1n estimate is 68 and the smudge is wrongly labeled as AAB, i.e. 3x68=204). It a smudge that is around 0.5, which indicates that it's either AB, AABB or AAABBB (or more, but that's unlikely I suppose).

If it's AB, then the haplod coverage could be 102, which is impossible given that there is one more peak ~50x in the kmer spectra. Hence it will be either AABB or AAABBB. You say that the plant is thought to be hexaploid, if it is case, the haploid coverage is ~34x, which is actually possible given the kmer histogram, it could be that 34x, 68x peaks are sort of blended together because the heterozygosity is not extremly high, but it is not conclusive, the coverage variation is really big. Was there any PCR step in the library prep? It almost looks like 10x library.

Finally, if you would like me to help to fix the smudgeplot, I would need the `coverage.tsv` file (the file you feed to smudgeplot) and perhaps it would be better if you could get it out of non-filtered, but trimmed data, is that possible? What L you have used? I would suggest L = 30 as the safest option for all the hypotheses we have.

**Comment by sivico26** [comment]
I just realize that I attached the wrong compressed file, ups!. Here are the coverages:

[`kmer_pairs_coverages_2.tsv`](https://drive.google.com/file/d/1ICFRmf372Z8bJYfRyRLr6aJ6TRTjCynC/view?usp=sharing), is in my drive since it is a little bit bulky for GitHub.

Just to clarify the filtering process, I followed this [procedure](https://gist.github.com/darencard/72ddd9e6c08aaff5ff64ca512a04a6dd) by Daren Card, without combining them in the final step. I did this first mapping the WGS reads to the chloroplast and then mapped the resulting unmapped reads to the mitochondrion, to keep the twice unmapped reads. 

Actually, I already did the smudges with the WGS data (without filtering) some time ago. Revisiting the results, they are actually pretty similar, but this seems to be correctly labeled in the x-axis (It does not agree with your above analysis though).

![ricinus_ploidy_smudgeplot_log10](https://user-images.githubusercontent.com/37163025/57854411-7aed4980-77e8-11e9-85c4-7fe37efb72d7.png)
![ricinus_ploidy_smudgeplot](https://user-images.githubusercontent.com/37163025/57854423-80e32a80-77e8-11e9-8fb5-17fbe8c9d29b.png)

L and U were set to 14 and 4000 in this case. In case they are useful, here I attach the hist and coverages files for this one too:

[`kmer_k21.hist.txt`](https://github.com/KamilSJaron/smudgeplot/files/3187089/kmer_k21.hist.txt)
[`kmer_pairs_WGS_coverages_2.tsv`](https://drive.google.com/file/d/1DQASkGsDHG1_IuDeKDh_ukosimT-Sq-R/view?usp=sharing)

Even though there is big coverage variation as you said, it was a normal lane HiSeq 4000 150x2 paired-end sequencing without any additional amplification (I say additional because Illumina is a clonal technology though).

I will follow your recommendation and rerun hetkmers in both cases using L = 30 and then I will be back to you.

Thanks for the help so far!

**Comment by KamilSJaron** [comment]
This genome is a piece of work ^^. The original wgs data look better then the filtered one. I am not sure about this procedure (I have not carefully checked the link, to be honest) and the data without it look nicer (less blended smudges), I think you must be filtering also some stuff that should not be filtered (because all mt and ct DNA should have super high coverage and it should not affect smudgeplot at all, if it does then I am suspecting something fishy).

Anyway, the best I could do in half an hour of playing with your data is using smudgeplot `v0.2.0`

```
smudgeplot plot kmer_pairs_WGS_coverages_2.tsv -o p_vol_WGS -n 47 -nbins 30
```

resulting in

![p_vol_WGS_smudgeplot](https://user-images.githubusercontent.com/8181573/57866594-e4c51d80-77ff-11e9-9d60-eadde66dd96a.png)

Then your genome would be a tetraploid, which would be quite quite consistent with genomescope. If you use the @tbenavi1 's [fork of genomescope](https://github.com/tbenavi1/genomescope) that is build also for polyploids you can fit model like this:

```
genomescope.R -i kmer_k21.hist.txt -o . -p 4 -n
```

![Pvol_plot](https://user-images.githubusercontent.com/8181573/57866686-09b99080-7800-11e9-97c7-85136ac6d787.png)

Why do I think that it's not a hexaploid?

If the big peak in genomescope represent 3n of your data, than your genome would be structured as AAABBB, which is very hard to imagine, you exect something like AAAABB or AABBCC so the chromosomal pairing is retained, AAABBB would have similar troubles with chromosome pairing as triploids. The if the big peak is 4n, then the smudgeplot would look very different.

Also this coverage sort of agrees with your assembly. Although, if the heterozygosity is realy ~10% between A and B, then I would suspect the assembly size to be way closer to dlouble of haploid genome size due to frequent separation of haplotypes in the assembly.

Finally, both smudgeplot and kmer hisogram represent what is in your sequencing reads and nothing else. The genomescope model is trying to find an interpretation of individual peaks (not always successfully) and smudgeplot is trying to correctly label individual smudges (also not always successfully). Even if the labels of smudges are misplaced, it does not change the fact that the underlying data is a representation of your sequencing reads/genome.

I already quite convinced about what am I looking at, but with L = 30 I think we will be finally able to cast a judgement.

**Comment by sivico26** [comment]
Wow! really nice plots you got there. I did not know about that fork of genomescope (which looks pretty interesting). You really convinced me as well about what is happening here, but just as final remark I will put the L=30 smudges I got:

**Note**: in both cases there were warnings about the high value of L, but I think it still got the distribution reflected.

### **Nuclear (filtered) Data**
![sacha_ploidy_auto_smudgeplot](https://user-images.githubusercontent.com/37163025/57918888-f78b3100-7897-11e9-99ee-8d82bc192064.png)
![sacha_ploidy_auto_smudgeplot_log10](https://user-images.githubusercontent.com/37163025/57918899-fbb74e80-7897-11e9-8d3a-51991c2c695e.png)

In this one the y-axis distribution is odd, probably due to the filtering that I definitively have to check again, but it kind of agrees with the one you got.

### **WGS Data**

![sacha_ploidy_auto_smudgeplot](https://user-images.githubusercontent.com/37163025/57919300-db3bc400-7898-11e9-96a4-d9d72841a1fe.png)
![sacha_ploidy_auto_smudgeplot_log10](https://user-images.githubusercontent.com/37163025/57919313-decf4b00-7898-11e9-9318-d971263a42f3.png)

But as you said, the whole data got better smudges, and this one is basically the same you produced, so I think that seals it.

As a curiosity, just as a final question: these ~5% of AAAABB kmers could suggest a few genes have several copies with one copy more frequent than others? Or it is more likely to come from a repetitive region? Certainly I would like to dig deeper in what the proportions of the smudges implies, somewhat similar as you suggest in the [idea](https://github.com/KamilSJaron/smudgeplot/wiki/idea) of smudgeplot and your final thoughts about the [strawberry example](https://github.com/KamilSJaron/smudgeplot/wiki/strawberry-tutorial).

I really thank your help through this process Kamil. I look forward the smudgeplot publication to cite the tool properly, please let us know when it is realeased!

**Comment by KamilSJaron** [comment]
Nice, you get the warning because 46 / 2 is < 30, but given all the exploration above, it's alright. I think we are done here :-).

> In this one the y-axis distribution is odd, probably due to the filtering that I definitively have to check again, but it kind of agrees with the one you got.

IMHO, the coverage "comb" looks totally like a bug somewhere. I would be worried to proceed with the filtered data, either figure out what has gone wrong or just assemble the original dataset (actually I am not sure why should mt/ctDNA be a problem).

> As a curiosity, just as a final question: these ~5% of AAAABB kmers could suggest a few genes have several copies with one copy more frequent than others?

Yes. However, smudgeplot considers only kmers that are "really close but not perfect", hence there is plenty of duplications/multiplications in the genome that won't be covered (like perfect duplications or duplications that are old). I would not spend too much time on direct interpretations of the paralog smudges, it just gives you a good prior what to expect in the assembly.

I spent some weeks trying to explore paralog smudges and how they can be used to understand better genome structure (the progress is partially recorded in [a GitHub Project](https://github.com/KamilSJaron/smudgeplot/projects/1)), but I could not make a good sense out of it. I explored a genome with tons of duplications, a bit like the diploid strawberry and the duplication smudges were mapping everywhere in the assembly, it was a mess and I did have time to give it an appropriate amount of attention hence I dropped it (for now).

> I really thank your help through this process Kamil. I look forward the smudgeplot publication to cite the tool properly, please let us know when it is realeased!

Happy to help. If you found our discussion helpful, I am always pleased to find my name in an acknowledgement ^^.

Closing for now, if anything, reopen this issue.
